ts/sbd-server/src/index.ts

```diff
@@ -0,0 +1,434 @@
+              ip,
+              nonce: true, // don't need to keep the actual nonce anymore
+              valid: true,
+            });
+
+            console.log(
+              'webSocketAuthenticated',
+              JSON.stringify({ pubKey: toB64Url(pubKey) }),
+            );
+          } else {
+            throw err(`invalid handshake message type ${msg.type()}`);
+          }
+        } else {
+          if (msg instanceof MsgNone) {
+            // no-op
+          } else if (msg instanceof MsgKeep) {
+            // keep alive
+          } else if (msg instanceof MsgForward) {
+            // extract the destination pubKey (slice does a copy)
+            const dest = msg.pubKey().slice(0);
+
+            // overwrite the destination pubKey with the source (our) pubKey
+            // the pubKey() function returns a reference, so editing it
+            // alters the message that will be sent
+            msg.pubKey().set(pubKey, 0);
+
+            const id = toB64Url(dest);
+            if (!this.queue[id]) {
+              this.queue[id] = [];
+            }
+            this.queue[id].push(msg.encoded());
+
+            if (!this.alarmLock) {
+              const alarm = await this.ctx.storage.getAlarm();
+              if (!alarm) {
+                this.ctx.storage.setAlarm(Date.now() + BATCH_DUR_MS);
+              }
+            }
+          } else {
+            throw err(`invalid post-handshake message type: ${msg.type()}`);
+          }
+        }
+      } catch (e: any) {
+        console.error('error', e.toString());
+        ws.close(4000 + (e.status || 500), e.toString());
+      }
+    });
+  }
+
+  /**
+   * The `webSocketMessage` handler above enqueues messages for delivery,
+   * then sets up an alarm to handle actually forwarding them. This ensures
+   * the messages are delivered in order without deadlocking two clients
+   * that happened to try to forward messages to each other at the same moment.
+   */
+  async alarm() {
+    const { shouldReturn, queue } = await this.ctx.blockConcurrencyWhile(
+      async () => {
+        if (this.alarmLock || Object.keys(this.queue).length === 0) {
+          return { shouldReturn: true, queue: {} };
+        }
+        this.alarmLock = true;
+        const queue = this.queue;
+        this.queue = {};
+        return { shouldReturn: false, queue };
+      },
+    );
+
+    if (shouldReturn) {
+      return;
+    }
+
+    // We cannot do the actual forwarding within a blockConcurrency because
+    // then if two peers try to send each other data at the same time it
+    // will deadlock. Hence all the complexity with the alarms and alarmLock.
+
+    for (const idName in queue) {
+      try {
+        const id = this.env.SIGNAL.idFromName(idName);
+        const stub = this.env.SIGNAL.get(id) as DurableObjectStub<DoSignal>;
+
+        await stub.forward(queue[idName]);
+      } catch (_e: any) {
+        // It is okay to get errors forwarding to peers, they may have
```

Comment:
log this?

Reply:
Mmm... I don't know how useful logging is going to be with cloudflare. Connecting an event listener restarts all durable objects, so I don't think we're going to want to do that in production. Moreover, sending messages to peers that are now offline is a common runtime situation that would flood the logs with messages that I don't know would be useful.

I think I'm actually going to need to remove all logging rather than add any more. We may need to brainstorm other ways of debugging cloudflare workers if it comes up that we have that need.